src/hotspot/share/opto/compile.cpp

```diff
@@ -4974,8 +4985,10 @@ void Compile::mark_unbalanced_boxes() {
           if (box != this_box) {
             box->set_unbalanced();
             this_box->set_unbalanced();
+            assert(this_box->is_marked_unbalanced(),"inconsistency");
           }
         }
+        assert(box->is_unbalanced() == box->is_marked_unbalanced(),"inconsistency");
```

Comment:
Could you even move this a scope out, i.e. a line down, so this check is even run with `alock->is_coarsened() == false`?

Reply:
Right. It will verify that we did not miss `is_unbalanced()` check when do EA or nested elimination to avoid them.

I will also move the assert at line 4984 under `(box != this_box)` to avoid duplication of check when `box` is the same.

Reply:
Moving assert down does not work because Nested lock elimination and EA can overwrite Coarsened status for BoxLock before we run `mark_unbalanced_boxes()`. Such BoxLock will not be marked as Unbalanced and assert will fail.
I would like to keep current assert in place but add an other assert when we change status of BoxLock. I am working on it.